src/typesystem.rs

```diff
@@ -559,9 +559,60 @@ fn get_rank(type_information: &DataTypeInformation) -> u32 {
     }
 }
 
-fn is_same_type_nature(ltype: &DataTypeInformation, rtype: &DataTypeInformation) -> bool {
-    (ltype.is_int() && ltype.is_int() == rtype.is_int())
-        || (ltype.is_float() && ltype.is_float() == rtype.is_float())
+pub fn is_same_type_nature(
+    ltype: &DataTypeInformation,
+    rtype: &DataTypeInformation,
+    index: &Index,
+) -> bool {
+    let ltype = find_inner_type_recursive(ltype, index);
+    let rtype = find_inner_type_recursive(rtype, index);
+    match ltype {
+        DataTypeInformation::Struct { name: lname, .. } => {
+            matches!(rtype, DataTypeInformation::Struct { name, ..} if name == lname)
+        }
+        DataTypeInformation::Array { name: lname, .. } => {
+            matches!(rtype, DataTypeInformation::Array { name, ..} if name == lname)
+        }
+        DataTypeInformation::Pointer {
+            inner_type_name: lname,
+            ..
+        } => {
+            matches!(rtype, DataTypeInformation::Pointer{ inner_type_name, ..} if inner_type_name == lname)
+        }
+        DataTypeInformation::Integer { .. } => matches!(rtype, DataTypeInformation::Integer { .. }),
+        DataTypeInformation::Enum { .. } => matches!(rtype, DataTypeInformation::Enum { .. }),
```

Comment:
the enum case does not look correct to me - i think this needs to be ltype == rtype